src/main/java/picard/util/SingleBarcodeDistanceMetric.java

```diff
@@ -0,0 +1,270 @@
+ * THE SOFTWARE.
+ */
+
+package picard.util;
+
+import htsjdk.samtools.util.SequenceUtil;
+
+import java.util.Arrays;
+
+/**
+ * A class for finding the distance between a single barcode and a barcode-read (with base qualities)
+ */
+public class SingleBarcodeDistanceMetric {
+
+    // The bases that the (comparison) barcode has
+    private final byte[] barcodeBases;
+
+    // The bases that the read has
+    private final byte[] readBases;
+
+    // the Read qualities
+    private final byte[] readQualities;
+
+    // The minimum base quality in the read that will be considered informative.
+    private final int minimumBaseQuality;
+
+    // a distance beyond which the user doesn't care how large the distance is
+    private final int maximalInterestingDistance;
+
+    public SingleBarcodeDistanceMetric(final byte[] barcodeBases,
+                                       final byte[] readBases,
+                                       final byte[] readQualities,
+                                       final int minimumBaseQuality,
+                                       final int maximalInterestingDistance) {
+        this.barcodeBases = barcodeBases;
+        this.readBases = readBases;
+        this.readQualities = readQualities;
+        this.minimumBaseQuality = minimumBaseQuality;
+        this.maximalInterestingDistance = maximalInterestingDistance;
+    }
+
+    public int hammingDistance() {
+        int numMismatches = 0;
+        for (int i = 0; i < barcodeBases.length && readBases.length > i && numMismatches <= maximalInterestingDistance; ++i) {
+            if (SequenceUtil.isNoCall(readBases[i])) {
+                continue;
+            }
+            if (!SequenceUtil.basesEqual(barcodeBases[i], readBases[i])) {
+                ++numMismatches;
+                continue;
+            }
+            // bases are equal but if quality is low we still penalize.
+            // TODO: is it actually useful to penalize barcodes in this way?
+            if (readQualities != null && readQualities[i] < minimumBaseQuality) {
+                ++numMismatches;
+            }
+        }
+        return numMismatches;
+    }
+
+    private static boolean anySmaller(final byte[] values, final int minValue) {
+        for (final byte value : values) {
+            if (value < minValue) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private byte[] maskSmaller() {
+        final byte[] maskedBases = Arrays.copyOf(readBases, readBases.length);
+
+        for (int i = 0; i < readQualities.length; i++) {
+            if (readQualities[i] < minimumBaseQuality) {
+                maskedBases[i] = '.';
+            }
+        }
+        return maskedBases;
+
+    }
+
+    private byte[] getMaskedReadBases() {
+        final byte[] maskedReadBases;
+        if (readQualities == null || !anySmaller(readQualities, minimumBaseQuality)) {
```

Comment:
the expectation is that in the vast majority of cases there will not be a smaller quality...and so this will prevent non-needed data copying. I'll smoosh the two functions together though as you suggest.

Reply:
I thin kI found a bug in the process :-/